enhancements/cloud-integration/ibm/service-endpoint-dynamic-override.md

```diff
@@ -0,0 +1,127 @@
+
+### User Stories
+
+* As an OpenShift cluster administrator, I want to update my cluster's current service endpoint configuration to point to new service endpoints.
+ 
+### Goals
+
+* Provide an official path for IBM deployed clusters to update the infrastructure object with service endpoint override(s) that will propagate to all dependent components without further user intervention.
+
+
+
+## Proposal
+
+To realize this enhancement:
+
+* Expand API definition to support defining services + endpoints within cloud provider spec for IBM
+* Modify CCCMO so that changes are reconciled from infrastructure spec to status for IBM cloud provider and cloud config
+* Evaluate IBM components with service endpoint support and how they will react/respond to post-install service endpoint changes.
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible managing an existing openshift custer deployed on IBM infrastructure.
+
+1. The cluster administrator wishes to use private IBM Cloud endpoints.
+2. The cluster administrator identifies the services that they wish to update (ie IAM and resource controller) and identifies the endpoints for these services.
+3. The cluster administrator updates the infrastructure object to contain a list of overrides where each element is the name of the service and the endpoint to use for that service. `oc edit infrastructure  -n default cluster`
+4. Once the service endpoint override update has been processed/reconciled, components can act on the change (if applicable) and use in future operations (note: may need to be restarted to pick up the change).
+
+**cccmo** is an operator responsible for watching updates to the infrastructure object and perforning updates once any value(s) are set. 
+
+1. The cccmo reconciliation loop observes that the IBMCloudPlatform spec within the infrastructure object has been set,
+2. The cccmo validates the endpoints and then updates the IBMCloudPlatformStatus and cloud config.
+
+
+### API Extensions
+
+* Extend `IBMCloudPlatformSpec` to contain service endpoint field that users may define as desired overrides.
+
+```
+type IBMCloudPlatformSpec struct {
+	// serviceEndpoints is a list of custom endpoints which will override the default
+	// service endpoints of an IBM Cloud service. These endpoints are consumed by
+	// components within the cluster to reach the respective IBM Cloud Services.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ServiceEndpoints []IBMCloudServiceEndpoint `json:"serviceEndpoints,omitempty"`
+}
+```
+
+### Topology Considerations
```

Comment:
do we ever need to consider a proxy scenario?

Reply:
What would happen if someone had clusterwideproxy configured?

Reply:
This isn't a particular scenario we wanted to support. Regarding clusterwide proxy; it isn't currently supported by IBM on ROKS and we would prefer to maintain some parity in this regard with IPI clusters. Was there a particular concern/reason you bring this up?

Reply:
It was my understanding that there's nothing that prevents a user from configuring the cluster wide proxy on an IPI cluster, and nothing in the documentation for the feature that says it's only supported on a subset of platforms. So we have to assume that at some point, a customer will try it on an IBM IPI cluster.

For other platforms, I believe the proxy implementation generates a list of excluded domains, eg like the AWS EC2 endpoints, so that calls to the cloud provider themselves aren't going through the proxy.

I would have expected something like this for IBM as well

Reply:
Sorry I should have been clearer - yes nothing prevents a user from enabling cluster-wide proxy on an IBM IPI. I want to come from the angle that should a user choose to enable the cluster-wide proxy they can also add the relevant endpoint overrides to the `noProxy` list and that they would be aware of that requirement. 

After digging through a bit, I have found IBM does not appear to have yet defined domains for exclusion should the proxy be enabled, unfortunately that todo has not been tackled.

Reply:
You might want to capture this in the doc, so it doesn't get left to the github comment ether. I'm ok with that being the suggested path, but having a record of that caveat here means we can then get it into product docs later